src/ReinforcementLearningZoo/src/algorithms/policy_gradient/maddpg.jl

```diff
@@ -0,0 +1,171 @@
+export MADDPGManager
+
+"""
+    MADDPGManager(; agents::Dict{<:Any, <:Agent}, args...)
+Multi-agent Deep Deterministic Policy Gradient(MADDPG) implemented in Julia. Here only works for simultaneous games whose action space is discrete.
+See the paper https://arxiv.org/abs/1706.02275 for more details.
+
+# Keyword arguments
+- `agents::Dict{<:Any, <:Agent{<:DDPGPolicy, <:AbstractTrajectory}}`, here each agent collects its own information. While updating the policy, each `critic` will assemble all agents' trajectory to update its own network.
+- `batch_size::Int`
+- `update_freq::Int`
+- `update_step::Int`, count the step.
+- `rng::AbstractRNG`.
+"""
+mutable struct MADDPGManager{P<:DDPGPolicy, T<:AbstractTrajectory} <: AbstractPolicy
+    agents::Dict{<:Any, <:Agent{<:P, <:T}}
+    batch_size::Int
+    update_freq::Int
+    update_step::Int
+    rng::AbstractRNG
+end
+
+# for simultaneous game with a discrete action space.
+function (π::MADDPGManager)(env::AbstractEnv)
+    while current_player(env) == chance_player(env)
+        env |> legal_action_space |> rand |> env
+    end
+    Dict((player, ceil(agent.policy(env))) for (player, agent) in π.agents)
+end
+
+function (π::MADDPGManager)(::PreEpisodeStage, ::AbstractEnv)
+    for (_, agent) in π.agents
+        if length(agent.trajectory) > 0
+            pop!(agent.trajectory[:state])
+            pop!(agent.trajectory[:action])
+            if haskey(agent.trajectory, :legal_actions_mask)
+                pop!(agent.trajectory[:legal_actions_mask])
+            end
+        end
+    end
+end
+
+function (π::MADDPGManager)(::PreActStage, env::AbstractEnv, actions)
+    # update each agent's trajectory
+    for (player, agent) in π.agents
+        push!(agent.trajectory[:state], state(env, player))
+        push!(agent.trajectory[:action], actions[player])
+        if haskey(agent.trajectory, :legal_actions_mask)
+            lasm = legal_action_space_mask(env, player)
+            push!(agent.trajectory[:legal_actions_mask], lasm)
+        end
+    end
+    
+    # update policy
+    update!(π)
+end
+
+function (π::MADDPGManager)(::PostActStage, env::AbstractEnv)
+    for (player, agent) in π.agents
+        push!(agent.trajectory[:reward], reward(env, player))
+        push!(agent.trajectory[:terminal], is_terminated(env))
+    end
+end
+
+function (π::MADDPGManager)(::PostEpisodeStage, env::AbstractEnv)
+    # collect state and dummy action to each agent's trajectory
+    for (player, agent) in π.agents
+        push!(agent.trajectory[:state], state(env, player))
+        push!(agent.trajectory[:action], rand(action_space(env)))
+        if haskey(agent.trajectory, :legal_actions_mask)
+            lasm = legal_action_space_mask(env, player)
+            push!(agent.trajectory[:legal_actions_mask], lasm)
+        end
+    end
+
+    # update policy
+    update!(π)
+end
```

Comment:
How about dispatching to the inner agent's corresponding methods?

Like calling `agent(stage, env, action)` in the `for` loop.

Reply:
Can you take a look at the `NamedPolicy` and see whether we can reuse existing code as much as possible? See also the `MultiAgentManager`